lib/posix/semaphore.c

```diff
@@ -1,11 +1,76 @@
 /*
  * Copyright (c) 2018 Intel Corporation
+ * Copyright (c) 2023 Meta
  *
  * SPDX-License-Identifier: Apache-2.0
  */
 
 #include <errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/posix/fcntl.h>
 #include <zephyr/posix/pthread.h>
+#include <zephyr/posix/semaphore.h>
+
+struct nsem_obj {
+	sys_snode_t snode;
+	sem_t sem;
+	atomic_t ref_count;
+	char *name;
```

Comment:
We could avoid malloc if allocated like this
```
char name[CONFIG_SEM_NAMELEN_MAX];
```
I think we have this pattern in many places in the code space.

Reply:
I'm gonna gather more comments for this one, doing so will allocate a fixed sized array regardless of the actual length of the name.

Alternatively we can allocate the buffer for the `nsem_obj` and the `name` array in one go and the total memory allocated will depend on the name length. Ideally we would be able to `k_realloc` the buffer once the sem is unlinked, however the `k_realloc` isn't available (#41151), so I opted for the current implementation as the name buffer can be freed separately upon `sem_unlink`.

Reply:
I would say a good follow-up PR would be to add dynamic allocation support.

The existing pattern that doesn't use malloc should be fine for filling the immediate gap. Thanks for your help ❤️